### PR TITLE
[12.0][MIG] Module stock_picking_invoicing

### PR DESCRIPTION
**Pull request by mbcosta**
Migration of module stock_picking_invoicing

cc @renatonlima @rvalyi @mileo

### CODE REVIEW

File: stock_picking_invoicing/wizards/stock_invoice_onshipping.py
Comment (celm1990) [inline stock_picking_invoicing/wizards/stock_invoice_onshipping.py]: Not is need journal type, purchase_refund and sale_refund not exist anymore

File: stock_picking_invoicing/wizards/stock_invoice_onshipping.py
Comment (celm1990) [inline stock_picking_invoicing/wizards/stock_invoice_onshipping.py]: Only keep 2 journals, 1 for sale and other for purchase, or 1 only and change domain

**Review by celm1990 — CHANGES_REQUESTED** [review]
@mbcosta what is status of this PR?. My comments in line

**Comment by mbcosta** [comment]
Hi @celm1990 , thank you by review. The status of this PR is waiting for reviews, I made the changes.

Hi @acsonefho , after this module was merged on version 10 I migrated it but as pointed by @celm1990 it's seems the journal type refund don't exist anymore since version 10, it'll be great if you can make a review to know your opnion about.

**Review by renatonlima — APPROVED** [review]

File: stock_picking_invoicing/models/account_invoice.py
Comment (celm1990) [inline stock_picking_invoicing/models/account_invoice.py:57]: I tested module and pass picking from invoice to refund is bad idea, on some countries refund have own picking for devolutions

File: stock_picking_invoicing/models/stock_picking.py
Comment (celm1990) [inline stock_picking_invoicing/models/stock_picking.py]: Same here, can cause conflict with module `stock_picking_invoice_link` https://github.com/OCA/stock-logistics-workflow/blob/12.0/stock_picking_invoice_link/models/stock_picking.py#L13

File: stock_picking_invoicing/models/account_invoice.py
Comment (celm1990) [inline stock_picking_invoicing/models/account_invoice.py]: Can cause conflict with module  `stock_picking_invoice_link` https://github.com/OCA/stock-logistics-workflow/blob/12.0/stock_picking_invoice_link/models/account_invoice.py#L13

Add dependency to module stock_picking_invoice_link can be possible???

**Review by celm1990 — CHANGES_REQUESTED** [review]

**Comment by acsonefho** [comment]
> Hi @celm1990 , thank you by review. The status of this PR is waiting for reviews, I made the changes.
> 
> Hi @acsonefho , after this module was merged on version 10 I migrated it but as pointed by @celm1990 it's seems the journal type refund don't exist anymore since version 10, it'll be great if you can make a review to know your opnion about.

Hi @mbcosta IMO everythink who is using refund journal could be removed (dedicated fields etc). Into a standard Odoo (checked on runbot), if you create a refund, you don't have to pick a refund journal because it doesn't exist anymore (as you said) and it use the journal set on the invoice. But I don't check if an OCA module re-introduce this refund journal.

**Comment by mbcosta** [comment]
Hi all

   @celm1990 I included the dependency of module stock_picking_invoice_link, good catch thanks,

   I included myself as contributor of the module. If it's needed to have more code done to include I remove it.

**Comment by rvalyi** [comment]
cc @acsonefho @sbidoul @mileo

**Comment by mileo** [comment]
Can you put 2invoice sum at stock.picking.type kanban? like 
image

I tested runbot several times but the invoices were created without lines

http://3402257-566-750700.runbot1-2.odoo-community.org/web#id=25&action=184&active_id=4&model=stock.picking&view_type=form&menu_id=100

WH/OUT/00010
WH/OUT/00011
WH/IN/00006

File: stock_picking_invoicing/wizards/stock_invoice_onshipping.py
Comment (mileo) [inline stock_picking_invoicing/wizards/stock_invoice_onshipping.py]: try to change:

lines = [] to  lines = [(5, 0, {})] 

and 

lines.append(line_values) to append((0, 0, line_values))

**Review by mileo — CHANGES_REQUESTED** [review]

File: stock_picking_invoicing/wizards/stock_invoice_onshipping.py
Comment (mileo) [reply stock_picking_invoicing/wizards/stock_invoice_onshipping.py]: and 

                    'invoice_line_ids': [(0, False, l) for l in lines],
to 

 invoice.write({ 'invoice_line_ids': lines})

**Review by mileo — CHANGES_REQUESTED** [review]
comments above

**Comment by mileo** [comment]
Can you please create a test, checking that the number of lines of the picking is the same of the invoice.

And another thing in Brazil invoice has a limt of lines numbers, can you implement a hook or a parameter that we can set that?

**Comment by mileo** [comment]
```
From 33537667fd43007d81e359bf928d702a452529d1 Mon Sep 17 00:00:00 2001
From: Luis Felipe Mileo <mileo@kmee.com.br>
Date: Fri, 30 Aug 2019 08:34:54 -0300
Subject: [PATCH] [PATCH] Separa os pickings ao faturar em lotes de 100 itens

Signed-off-by: Luis Felipe Mileo <mileo@kmee.com.br>
---
 addons/stock_account/stock.py | 103 ++++++++++++++++++++--------------
 1 file changed, 61 insertions(+), 42 deletions(-)

diff --git a/addons/stock_account/stock.py b/addons/stock_account/stock.py
index 48731b49a7c..24a1c2b3e97 100644
--- a/addons/stock_account/stock.py
+++ b/addons/stock_account/stock.py
@@ -21,6 +21,22 @@
 
 from openerp.osv import fields, osv
 
+
+def chunk(iterable, chunk_size):
+    """Generate sequences of `chunk_size` elements from `iterable`."""
+    iterable = iter(iterable)
+    while True:
+        chunk = []
+        try:
+            for _ in range(chunk_size):
+                chunk.append(iterable.next())
+            yield chunk
+        except StopIteration:
+            if chunk:
+                yield chunk
+            break
+
+
 class stock_location_path(osv.osv):
     _inherit = "stock.location.path"
     _columns = {
@@ -279,7 +295,7 @@ class stock_picking(osv.osv):
     def _get_invoice_vals(self, cr, uid, key, inv_type, journal_id, move, context=None):
         if context is None:
             context = {}
-        partner, currency_id, company_id = key
+        partner, currency_id, company_id, idx = key
         if inv_type in ('out_invoice', 'out_refund'):
             account_id = partner.property_account_receivable.id
             payment_term = partner.property_payment_term.id or False
@@ -306,48 +322,51 @@ class stock_picking(osv.osv):
         invoices = {}
         is_extra_move, extra_move_tax = move_obj._get_moves_taxes(cr, uid, moves, inv_type, context=context)
         product_price_unit = {}
-        for move in moves:
-            company = move.company_id
-            origin = move.picking_id.name
-            partner, user_id, currency_id = move_obj._get_master_data(cr, uid, move, company, context=context)
-
-            key = (partner, currency_id, company.id)
-            invoice_vals = self._get_invoice_vals(cr, uid, key, inv_type, journal_id, move, context=context)
 
-            if key not in invoices:
-                # Get account and payment terms
-                invoice_id = self._create_invoice_from_picking(cr, uid, move.picking_id, invoice_vals, context=context)
-                invoices[key] = invoice_id
-            else:
-                invoice = invoice_obj.browse(cr, uid, invoices[key], context=context)
-                merge_vals = {}
-                if not invoice.origin or invoice_vals['origin'] not in invoice.origin.split(', '):
-                    invoice_origin = filter(None, [invoice.origin, invoice_vals['origin']])
-                    merge_vals['origin'] = ', '.join(invoice_origin)
-                if invoice_vals.get('name', False) and (not invoice.name or invoice_vals['name'] not in invoice.name.split(', ')):
-                    invoice_name = filter(None, [invoice.name, invoice_vals['name']])
-                    merge_vals['name'] = ', '.join(invoice_name)
-                if merge_vals:
-                    invoice.write(merge_vals)
-            invoice_line_vals = move_obj._get_invoice_line_vals(cr, uid, move, partner, inv_type, context=dict(context, fp_id=invoice_vals.get('fiscal_position', False)))
-            invoice_line_vals['invoice_id'] = invoices[key]
-            invoice_line_vals['origin'] = origin
-            if not is_extra_move[move.id]:
-                product_price_unit[invoice_line_vals['product_id'], invoice_line_vals['uos_id']] = invoice_line_vals['price_unit']
-            if is_extra_move[move.id] and (invoice_line_vals['product_id'], invoice_line_vals['uos_id']) in product_price_unit:
-                invoice_line_vals['price_unit'] = product_price_unit[invoice_line_vals['product_id'], invoice_line_vals['uos_id']]
-            if is_extra_move[move.id]:
-                desc = (inv_type in ('out_invoice', 'out_refund') and move.product_id.product_tmpl_id.description_sale) or \
-                    (inv_type in ('in_invoice','in_refund') and move.product_id.product_tmpl_id.description_purchase)
-                invoice_line_vals['name'] += ' ' + desc if desc else ''
-                if extra_move_tax[move.picking_id, move.product_id]:
-                    invoice_line_vals['invoice_line_tax_id'] = extra_move_tax[move.picking_id, move.product_id]
-                #the default product taxes
-                elif (0, move.product_id) in extra_move_tax:
-                    invoice_line_vals['invoice_line_tax_id'] = extra_move_tax[0, move.product_id]
-
-            move_obj._create_invoice_line_from_vals(cr, uid, move, invoice_line_vals, context=context)
-            move_obj.write(cr, uid, move.id, {'invoice_state': 'invoiced'}, context=context)
+        parts = chunk(moves, 100)
+        for idx, item in enumerate(parts):
+            for move in item:
+                company = move.company_id
+                origin = move.picking_id.name
+                partner, user_id, currency_id = move_obj._get_master_data(cr, uid, move, company, context=context)
+
+                key = (partner, currency_id, company.id, idx)
+                invoice_vals = self._get_invoice_vals(cr, uid, key, inv_type, journal_id, move, context=context)
+
+                if key not in invoices:
+                    # Get account and payment terms
+                    invoice_id = self._create_invoice_from_picking(cr, uid, move.picking_id, invoice_vals, context=context)
+                    invoices[key] = invoice_id
+                else:
+                    invoice = invoice_obj.browse(cr, uid, invoices[key], context=context)
+                    merge_vals = {}
+                    if not invoice.origin or invoice_vals['origin'] not in invoice.origin.split(', '):
+                        invoice_origin = filter(None, [invoice.origin, invoice_vals['origin']])
+                        merge_vals['origin'] = ', '.join(invoice_origin)
+                    if invoice_vals.get('name', False) and (not invoice.name or invoice_vals['name'] not in invoice.name.split(', ')):
+                        invoice_name = filter(None, [invoice.name, invoice_vals['name']])
+                        merge_vals['name'] = ', '.join(invoice_name)
+                    if merge_vals:
+                        invoice.write(merge_vals)
+                invoice_line_vals = move_obj._get_invoice_line_vals(cr, uid, move, partner, inv_type, context=dict(context, fp_id=invoice_vals.get('fiscal_position', False)))
+                invoice_line_vals['invoice_id'] = invoices[key]
+                invoice_line_vals['origin'] = origin
+                if not is_extra_move[move.id]:
+                    product_price_unit[invoice_line_vals['product_id'], invoice_line_vals['uos_id']] = invoice_line_vals['price_unit']
+                if is_extra_move[move.id] and (invoice_line_vals['product_id'], invoice_line_vals['uos_id']) in product_price_unit:
+                    invoice_line_vals['price_unit'] = product_price_unit[invoice_line_vals['product_id'], invoice_line_vals['uos_id']]
+                if is_extra_move[move.id]:
+                    desc = (inv_type in ('out_invoice', 'out_refund') and move.product_id.product_tmpl_id.description_sale) or \
+                        (inv_type in ('in_invoice','in_refund') and move.product_id.product_tmpl_id.description_purchase)
+                    invoice_line_vals['name'] += ' ' + desc if desc else ''
+                    if extra_move_tax[move.picking_id, move.product_id]:
+                        invoice_line_vals['invoice_line_tax_id'] = extra_move_tax[move.picking_id, move.product_id]
+                    #the default product taxes
+                    elif (0, move.product_id) in extra_move_tax:
+                        invoice_line_vals['invoice_line_tax_id'] = extra_move_tax[0, move.product_id]
+
+                move_obj._create_invoice_line_from_vals(cr, uid, move, invoice_line_vals, context=context)
+                move_obj.write(cr, uid, move.id, {'invoice_state': 'invoiced'}, context=context)
 
         invoice_obj.button_compute(cr, uid, invoices.values(), context=context, set_total=(inv_type in ('in_invoice', 'in_refund')))
         return invoices.values()
-- 
2.22.0

```

@mbcosta  this is a path that i need to do in one customer.

You can create a hook that we can change the number "100" easily....

**Comment by mbcosta** [comment]
Hi @mileo thank you for review. I included in the kanban view the pickings in invoice state 2binvoiced ( good idea ) and the tests to check invoice lines. The error you find in runbot occur because the field invoice_state of stock.move was send with none instead 2binvoiced.

 About 'And another thing in Brazil invoice has a limt of lines numbers, can you implement a hook or a parameter that we can set that?' my first  thought is it should made in the Brazilian Localization project ( I think we should open an issue about it there ), but I like to hear the opinion of the others ( cc @renatonlima @rvalyi @acsonefho @sbidoul @celm1990 )  if we should implement here or there or even create a new module.

**Comment by renatonlima** [comment]
Hi @mileo,

I think we can override this method in l10n-brazil:  https://github.com/OCA/account-invoicing/pull/566/files#diff-a74949d63b43263dc7f988cd2bd1a65aR249

**Review by renatonlima — APPROVED** [review]

**Comment by mileo** [comment]
@renatonlima we need to ungroup stock.move in many invoices. 

Can you tell me how we can do this by overwriting this method?

**Comment by rvalyi** [comment]
@mileo @renatonlima I agree the _get_picking_key is not enough to split invoices from the same existing picking (it can eventually prevent grouping pickings together at most). I'm not sure what kind of hook we could have. @mileo would you have a hook to suggest?

**Comment by mileo** [comment]
> @mileo @renatonlima I agree the _get_picking_key is not enough to split invoices from the same existing picking (it can eventually prevent grouping pickings together at most). I'm not sure what kind of hook we could have. @mileo would you have a hook to suggest?

Hello, I suggest something like this .... It would make it much easier to change the amount of items in a custom module:

```
def chunk(iterable, chunk_size):
    """Generate sequences of `chunk_size` elements from `iterable`."""
    if not chunk_size:
        return iterable
    iterable = iter(iterable)
    while True:
        chunk = []
        try:
            for _ in range(chunk_size):
                chunk.append(iterable.next())
            yield chunk
        except StopIteration:
            if chunk:
                yield chunk
            break

class StockInvoiceOnshipping(models.TransientModel):
    _name = 'stock.invoice.onshipping'
    _description = "Stock Invoice Onshipping"

    [...]

    def _chunk_size()
         return False

    @api.multi
    def _action_generate_invoices(self):
        """
        Action to generate invoices based on pickings
        :return: account.invoice recordset
        """
        pickings = self._load_pickings()
        company = pickings.mapped("company_id")
        if company and company != self.env.user.company_id:
            raise UserError(_("All pickings are not related to your company!"))
        pick_list = self._group_pickings(pickings)
        invoices = self.env['account.invoice'].browse()
        for pickings in pick_list:
            invoice_values = self._build_invoice_values_from_pickings(pickings)
            all_moves = pickings.mapped("move_lines")
    
            parts = chunk(all_moves, self._chunk_size())
            for idx, item in enumerate(parts):
                for moves in item:
                    invoice = self.env['account.invoice'].create(invoice_values)
                    moves_list = self._group_moves(moves)
                    lines = []
                    for moves in moves_list:
                        line_values = self._get_invoice_line_values(moves, invoice)
                        if line_values:
                            lines.append(line_values)
                    if lines:
                        invoice.write({
                            'invoice_line_ids': [(0, False, l) for l in lines],
                        })
                    invoice._onchange_invoice_line_ids()
                    invoices |= invoice
        # Update the state on pickings related to new invoices only
        self._update_picking_invoice_status(invoices.mapped("picking_ids"))
        return invoices
```

**Comment by renatonlima** [comment]
@mileo,

I think it's better to do this migration merge and then do a PR with this feature, what do you think?

**Comment by mileo** [comment]
I think we better solve this question now. But if @mbcosta  can't work on it I can allocate someone from our team to finish it on monday. 

cc @kmee

File: stock_picking_invoicing/models/stock_picking_type.py
Comment (mileo) [inline stock_picking_invoicing/models/stock_picking_type.py]: This domain must be only:
```
('state', '!=', 'cancel'),
```
The user must see the pickings that are already done, and ready to invoice; Or may be the 2 pick too, but 2binvoiced;

And the view domain must be the same, because you se for example 3 in the dashboard, and when you open you have 10.

File: stock_picking_invoicing/views/stock_picking.xml
Comment (mileo) [inline stock_picking_invoicing/views/stock_picking.xml]: This button is not being displayed

File: stock_picking_invoicing/views/stock_picking.xml
Comment (mileo) [inline stock_picking_invoicing/views/stock_picking.xml:69]: 1. Can you put this field above the picking type?
2. You can't change the state of this field if the picking is not draft;
3. You must ensure the sync between this field and the stock.moves;
4. Should it be allowed some itens that will be invoiced and others not in the same picking?

**That's why I detected the bug of having an invoice without lines;**

File: stock_picking_invoicing/views/stock_picking_type.xml
Comment (mileo) [inline stock_picking_invoicing/views/stock_picking_type.xml:12]: Please ensure that this domain is the same of the count.

File: stock_picking_invoicing/tests/test_picking_invoicing.py
Comment (mileo) [inline stock_picking_invoicing/tests/test_picking_invoicing.py:21]: `self.pick_type_in_count = self.pick_type_in.count_picking_2binvoiced`

Comment (mileo) [reply stock_picking_invoicing/tests/test_picking_invoicing.py:21]: and something like that in the tests above:

```

        self.assertEqual(
            self.pick_type_in_count + 1,
            self.pick_type_in.count_picking_2binvoiced
        )
```

**Review by mileo — CHANGES_REQUESTED** [review]

**Comment by mileo** [comment]
@mbcosta @renatonlima @rvalyi PR at akretion with chunk https://github.com/akretion/account-invoicing/pull/8

**Comment by mileo** [comment]
Hello, I made two new improvements at https://github.com/akretion/account-invoicing/pull/8:

1. Improve invoice creation performance;
2. Some small code refactor to allow to create a module that uses queue_job to create bath invoices

**Comment by mbcosta** [comment]
Hi @mileo , I made the changes you ask for, about:

You can't change the state of this field if the picking is not draft;
You must ensure the sync between this field and the stock.moves;
Should it be allowed some itens that will be invoiced and others not in the same picking?

I think the three questions are related, by the actual code the answer is yes, I don't see any problem to keep the actual desing or change it, so what is your arguments to made this changes ?  Why the user can't set stock picking in state Done To Be Invoiced ? Why we can't have in the same picking moves with invoice_state 2binvoiced, invoiced and none ?

About your PR I agree with @rvalyi wrote there, we should just included here methods to be overridden in other modules. The implementation of an limit to the numbers of the invoice lines generated is very specific thing and should be in brazilian localization or other module/repo.

**Comment by mileo** [comment]
> Hi @mileo , I made the changes you ask for, about:
> 
> You can't change the state of this field if the picking is not draft;
> You must ensure the sync between this field and the stock.moves;
> Should it be allowed some itens that will be invoiced and others not in the same picking?
> 
> I think the three questions are related, by the actual code the answer is yes, I don't see any problem to keep the actual desing or change it, so what is your arguments to made this changes ? Why the user can't set stock picking in state Done To Be Invoiced ? Why we can't have in the same picking moves with invoice_state 2binvoiced, invoiced and none ?
> 
> About your PR I agree with @rvalyi wrote there, we should just included here methods to be overridden in other modules. The implementation of an limit to the numbers of the invoice lines generated is very specific thing and should be in brazilian localization or other module/repo.

I do not agree but I redo the improvement commits without the chunk.

Please cherry pick of it:

Ungroup hook:
https://github.com/akretion/account-invoicing/pull/8/commits/55dddf35cd4be1e7badcb35a8d2aac462cd6dde1

[MOV] _update_picking_invoice_status to allow queue_job calls
https://github.com/akretion/account-invoicing/pull/8/commits/021b0f25dc26c3bd40d6878dca890108aab14860

Improve invoice create perfomance:
https://github.com/akretion/account-invoicing/pull/8/commits/cbf2cb6e09479c544b62f980fdd01fdf0996f9da

**Comment by mbcosta** [comment]
Hi @mileo thanks, to include an contributor you should put this information on readme/CONTRIBUTORS.rst file and then use the oca-gen-addon-readme command to generated README.rst automatically https://github.com/oca/maintainer-tools as I wrote in your PR.

**Comment by mileo** [comment]
> Hi @mileo thanks, to include an contributor you should put this information on readme/CONTRIBUTORS.rst file and then use the oca-gen-addon-readme command to generated README.rst automatically https://github.com/oca/maintainer-tools as I wrote in your PR.

As done in commit https://github.com/akretion/account-invoicing/commit/cbf2cb6e09479c544b62f980fdd01fdf0996f9da?

image

**Comment by mbcosta** [comment]
@mileo I think is missing run oca-gen-addon-readme command to filled the session Contributor on README.rst.

**Comment by mileo** [comment]
This is automatically done by the ocabot when maintainers use the command "ocabot merge"

Please see https://odoo-community.org/blog/the-oca-blog-1/post/the-oca-github-bot-can-now-merge-pull-requests-92 or https://github.com/OCA/oca-github-bot/blob/7848453162b4042d3206ecfac0f5fb4bf759f583/src/oca_github_bot/tasks/merge_bot.py#L98

File: stock_picking_invoicing/tests/__init__.py
Comment (renatonlima) [inline stock_picking_invoicing/tests/__init__.py]: Please remove this python comment line

File: stock_picking_invoicing/tests/__init__.py
Comment (renatonlima) [inline stock_picking_invoicing/tests/__init__.py:4]: Add line after python comment

File: stock_picking_invoicing/tests/test_picking_invoicing.py
Comment (renatonlima) [inline stock_picking_invoicing/tests/test_picking_invoicing.py]: Please remove this python comment line

**Review by renatonlima — CHANGES_REQUESTED** [review]

File: stock_picking_invoicing/__init__.py
Comment (renatonlima) [inline stock_picking_invoicing/__init__.py]: Is it should be in a migration script?

**Comment by mbcosta** [comment]
Hi @mileo I tried to include what you did but your code break the actual code, it's fail in the test and  the wizard to create invoice starts to create a invoice per stock_moves, please check out.

**Comment by mileo** [comment]
@mbcosta sorry, cherry pick this fix: https://github.com/akretion/account-invoicing/pull/8/commits/f4c03a4ad47c687d9bd5eb0770dcd7a6ecea1a90

**Comment by mbcosta** [comment]
@mileo your changes was merged, thanks

**Review by rvalyi — APPROVED** [review]
Hello kudos for the work @mbcosta !
I reviewed the code and tested and I looks good to me.
A minor improvement I can see is in wizards/stock_invoice_onshipping.py#action_generate:
indeed, when there is only 1 invoice (the usual case), we could open the invoice form view directly instead of the tree view. This is not a regression against the 10.0 version so I'm not blocking the PR for this but this could be a nice improvement.

**Comment by mbcosta** [comment]
Hi @rvalyi thanks for review, I made the improvement

**Comment by mileo** [comment]
@mbcosta Trying to review this module again, i can't confirm the invoice.

Having this error:

image

I also checked the https://coveralls.io/builds/28355247/source?filename=stock_picking_invoicing/models/stock_move.py#L21

Trying to create the same invoice manually, the taxes are automatically filled....

Could you please check this issue?

**Comment by mbcosta** [comment]
Hi @mileo thanks again for your review. The only problem I found about taxes was the code just load sale taxes ( field taxes_id of product.template ) the purchase taxes ( supplier_taxes_id of product.template ) was not ( maybe it's your test case ), I fixed it and included the check if the invoice lines created has taxes.

**Review by renatonlima — CHANGES_REQUESTED** [review]

**Comment by mileo** [comment]
@mbcosta please take a look at this PR: https://github.com/akretion/account-invoicing/pull/9

That fixes the invoice taxes.

**Comment by mbcosta** [comment]
Hi @renatonlima @mileo thanks for your review, I did and merge the changes

**Review by renatonlima — APPROVED** [review]

**Review by mileo — APPROVED** [review]

**Comment by mileo** [comment]
UP

**Pull request merged**
